### PR TITLE
feat: card atom

### PR DESCRIPTION
**Pull request by guilhermevrs**
### Objective

The idea here is to have a base card component and to extend it for the real cards.
This would enable for example to extend the card-ui to have:

- Normal playing cards
- Tarot playing cards
- etc

This PR includes then:

- Update of testing dependencies
- Fixed issue with react-scripts and Jest. Check https://github.com/testing-library/dom-testing-library/issues/477
- Utils for components

### Missing actions
- [x] Need the linting / formating
- [x] Need the CI
- [x] Squash once finish

Closes #16

### CODE REVIEW

File: src/utils/component-utils.tsx [hunk 1]
```diff
@@ -0,0 +1,13 @@
+import { BaseProps } from "../models/BaseProps";
+
+/**
+ * Returns the props to be spreaded at component container level
+ * @param param0 component base props
+ */
+export function getContainerOps({ testId }: BaseProps) {
```
Comment (tib-tib) [inline src/utils/component-utils.tsx:7]: I don't really get what does this function by reading at the name, and even in general I don't really understand its purpose, could you explain a bit more?

File: src/models/BaseProps.tsx [hunk 2]
```diff
@@ -0,0 +1,7 @@
+/**
+ * Base properties for component
+ */
+export interface BaseProps {
+    /** testable  id */
+    testId?: string
+}
```
Comment (tib-tib) [inline src/models/BaseProps.tsx:7]: why is this needed?

File: src/components/molecules/BaseCard/BaseCard.tsx [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+  }
+});
+
+/**
+ * Base card molecule
+ * @param props Card props
+ */
+export const BaseCard = (props: CardProps = {}) => {
+  const classes = useStyles();
+  const [cardState, setCardState] = useState<CardState>({visibleFace: 'front'});
```
Comment (tib-tib) [inline src/components/molecules/BaseCard/BaseCard.tsx:82]: Here I think it could be simpler, we could represent the state of a card with a boolean, like:

```
const [isFrontVisible, setIsFrontVisible] = useState(false);
```

I don't realy like having a string describing which face is visible

File: src/utils/component-utils.tsx (hunk 1 above)
Comment (guilhermevrs) [reply src/utils/component-utils.tsx:7]: I removed.
I guess we would need to some kind of base properties for the component containers, at least so we can target them at test. But with the lint forcing me to include role for example, we can manage it without for now.

File: src/models/BaseProps.tsx (hunk 2 above)
Comment (guilhermevrs) [reply src/models/BaseProps.tsx:7]: Same as before

File: src/components/molecules/BaseCard/BaseCard.tsx (hunk 3 above)
Comment (guilhermevrs) [reply src/components/molecules/BaseCard/BaseCard.tsx:82]: Now here is indeed a part where I couldn't make my mind.
Should we have several `useState` or a single one  holding the whole component? I am ok having it several, but I wasn't sure what is the pros / cons of this.

Comment (guilhermevrs) [reply src/components/molecules/BaseCard/BaseCard.tsx:82]: Now, for the second point.
This is not a simple string, it's a string type. Compiler will force either `front` or `back`. Anything different, it will break.

I guess this makes the code more readable (`if visibleFace === 'back'` instead of `is isFrontVisible`)

Any arguments on why you don't like the string as describing which face is visible?

File: src/index.tsx [hunk 4]
```diff
@@ -8,7 +8,7 @@ ReactDOM.render(
   <React.StrictMode>
     <App />
   </React.StrictMode>,
-  document.getElementById('root')
+  document.getElementById('root'),
```
Comment (guilhermevrs) [inline src/index.tsx:11]: Is this coming from prettier, eslint or my editor?
It's weird to me to see the last item with a comma

File: src/setupTests.ts [hunk 5]
```diff
@@ -1,5 +1,6 @@
+/* eslint-disable import/no-extraneous-dependencies */
```
Comment (guilhermevrs) [inline src/setupTests.ts:1]: I included here because test dependencies, for me, is always a devDependency.
So need to tell eslint to not consider this file when doing the above check

**Comment by guilhermevrs** [comment]
image

File: src/components/atoms/BaseCard/BaseCard.tsx [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    transform: 'rotateY(0deg)',
+  },
+  cardBack: {
+    extend: 'cardFace',
+    transform: 'rotateY(180deg)',
+  },
+});
+
+/**
+ * Base card molecule
```
Comment (tib-tib) [inline src/components/atoms/BaseCard/BaseCard.tsx:65]: * Base card

In general, don't put the type of component (atom, molecule, etc) in the jsdoc, since it might change and it doesn't add any value

File: src/index.tsx (hunk 4 above)
Comment (tib-tib) [reply src/index.tsx:11]: It's because in prettier it was configured `"trailingComma": "all"` as you suggested, but I think we should use `"trailingComma": "es5"` (the default value) to avoid that

File: src/components/atoms/BaseCard/BaseCard.tsx [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+    backfaceVisibility: 'hidden',
+    position: 'absolute',
+    top: 0,
+    left: 0,
+    color: 'cornSilk',
+    textAlign: 'center',
+    font: `3em/240px 'Helvetica Neue', Helvetica, sans-serif`,
+    boxShadow: '-5px 5px 5px #aaa',
+    transition: '0.6s',
+  },
```
Comment (tib-tib) [inline src/components/atoms/BaseCard/BaseCard.tsx:30]: do we need all this styling if the card is an svg?

File: src/components/atoms/BaseCard/BaseCard.tsx [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+  card: ({ width, height }) => ({
+    width: width || 600,
+    height: height || 400,
+    perspective: 1000,
+    cursor: 'pointer',
+    '&$backVisible $cardFlipper': { transform: 'rotateY(180deg)' },
+    '&$backVisible $cardFront': { opacity: 0 },
+    '&$backVisible $cardBack': { opacity: 1 },
+    '&$frontVisible $cardFront': { opacity: 1 },
+    '&$frontVisible $cardBack': { opacity: 0 },
```
Comment (tib-tib) [inline src/components/atoms/BaseCard/BaseCard.tsx:44]: I don't really like this part, feels kind of duplicated, can't we simplify it?

File: src/components/atoms/BaseCard/BaseCard.tsx [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+ * @param props Card props
+ */
+export const BaseCard: React.FC<BaseCardProps> = ({
+  frontFace,
+  backFace,
+  width = undefined,
+  height = undefined,
+} = {}) => {
+  const { backVisible, frontVisible, card, cardFlipper, cardFront, cardBack } = useStyles({ width, height });
+  const [visibleFace, setVisibleFace] = useState<'front' | 'back'>('front');
```
Comment (tib-tib) [inline src/components/atoms/BaseCard/BaseCard.tsx:75]: so regarding my comment from before it's ok to keep it as string I guess but the problem to me is that you hardcode them. It would be better if you'd use a const for that like 
``` 
const frontFace = 'front';
const backFace = 'back';
type VisibleFace = frontFace | backFace;
```
and then later on

```
const [visibleFace, setVisibleFace] = useState<VisibleFace>(frontFace);
```

and so on, problem here is that if at some point we decide to rename the 'front' string into 'frontFace' for instance, you'd have to change it everywhere in the code instead of changing it in one place only.

**Comment by tib-tib** [comment]
In my mind I was picturing the Card component to be much simpler, something like:

```
return (
  <img src={isFront ? frontFace : backFace } />
)
```
, with frontFace and backFace being two svg urls. And that was it haha!

I don't think having a `generateSvg` method makes sense, because depending on the kind of cards we would have (Tarot, UNO, whatever) we might not have the same way of retrieving/generating them.
At some point we might even get a deck of cards from the backend and it would be a list of something like Card: { value: 'ace', color: 'spades', url: 'url-of-my-svg-card' }.

Anyway, I think what you did is perfectly fine, just wanted to express my thoughts :)

File: src/index.tsx (hunk 4 above)
Comment (guilhermevrs) [reply src/index.tsx:11]: Indeed, I don't remember from where I got this prettier config, but trailingComma all is very weird.
es5 seems fine indeed 👍

File: src/components/atoms/BaseCard/BaseCard.tsx (hunk 6 above)
Comment (guilhermevrs) [reply src/components/atoms/BaseCard/BaseCard.tsx:65]: Indeed, very much agree

File: src/components/atoms/BaseCard/BaseCard.tsx (hunk 7 above)
Comment (guilhermevrs) [reply src/components/atoms/BaseCard/BaseCard.tsx:30]: Indeed all that part for font and boxShadow we can remove it, leaving only position only

File: src/components/atoms/BaseCard/BaseCard.tsx (hunk 8 above)
Comment (guilhermevrs) [reply src/components/atoms/BaseCard/BaseCard.tsx:44]: I don't like either!
But to be honest, I have tried to find a way of fixing it at react-jss but no clue..

It's kind of weird, I am still trying to find a way that pleases me. What do you think about:
```typescript
'&$backVisible $cardFront, &$frontVisible $cardBack': { opacity: 0 },
'&$backVisible $cardBack, &$frontVisible $cardFront': { opacity: 1 },
```
?

File: src/components/atoms/BaseCard/BaseCard.tsx (hunk 9 above)
Comment (guilhermevrs) [reply src/components/atoms/BaseCard/BaseCard.tsx:75]: So, by your suggestion, it would become:

```typescript
const frontLabel = 'front';
const backLabel = 'back';
type VisibleFace = typeof frontLabel | typeof backLabel;

export const BaseCard: React.FC<BaseCardProps> = ({
  frontFace,
  backFace,
  width = undefined,
  height = undefined,
} = {}) => {
       const [visibleFace, setVisibleFace] = useState<VisibleFace>(frontFace);
       const turnCard = (): void => setVisibleFace(visibleFace === frontLabel ? backLabel : frontLabel);

...
```
Don't you feel it a bit unnatural?
I mean, in general I am totally inline for not hardcoding because of two things:

1 - If you need to rewrite on several places
2 - if you forget to replace somewhere, you may break the app without knowing (like in JS without type checking)

In here we are talking about:
1 - a single file with less then 100 lines
2 - if you forget to replace somewhere, your build will fail and will tell you exact the lines where you should be updating.

I don't know, never seen this kind of approach in typescript
We can discuss over our talk later

**Comment by guilhermevrs** [comment]
Well, what I was thinking for the PlayingCard is to be an extension of BaseCard.
So BaseCard could be any card, since the only thing you can do is to place it somewhere and turn it via click.

I called PlayingCard because I was referring to a "normal playing card". I would imagine a UnoCard or TarotCard if it's the thing. Specially because the type of entry would be different (I mean, you wouldn't instantiate the UnoCard with JokerBlack).

Now about using the `img` tag. You can only do it if you have an SVG separated for each card. In this one, it's a SVG Sprite (so a single download can get you all the cards already). Since it's an SVG sprite, you can't simply use the `<img src=` I guess.
Moreover, using img src blocks you of using SVG inline attributes (like `fill` for example).

At server level, indeed, you could receive the data (like the SVG url). But then would that be a single SVG or an SVG sprite? 
If it's a single, we could have a mode for loading the card from single SVG (in that case, two svgs, one for the front and one for the back).

Again, let's discuss during our call

File: src/components/molecules/BaseCard/BaseCard.tsx (hunk 3 above)
Comment (guilhermevrs) [reply src/components/molecules/BaseCard/BaseCard.tsx]: In the end, after moving to `classnames`, having the boolean is  easier. I have changed

**Comment by guilhermevrs** [comment]
Closes #3
